File: public/cpp/053-house-robber-ii.cpp
/*
 * LeetCode Problem #213: House Robber II
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/house-robber-ii/
 *
 * Problem Statement:
 * You are a professional robber planning to rob houses along a street. Each house
 * has a certain amount of money stashed. All houses are arranged in a circle, which
 * means the first house is the neighbor of the last one. Adjacent houses have
 * security systems connected — robbing two adjacent houses triggers an alarm.
 *
 * Given an integer array nums representing the amount of money of each house,
 * return the maximum amount of money you can rob tonight without alerting the police.
 *
 * Example 1:
 * Input: nums = [2,3,2]
 * Output: 3
 * Explanation: You cannot rob house 1 (2) and house 3 (2) because they are adjacent.
 *
 * Example 2:
 * Input: nums = [1,2,3,1]
 * Output: 4
 * Explanation: Rob house 1 (1) and house 3 (3). Total = 4.
 *
 * Example 3:
 * Input: nums = [1,2,3]
 * Output: 3
 *
 * Constraints:
 * - 1 <= nums.length <= 100
 * - 0 <= nums[i] <= 1000
 */

#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;

/*
 * APPROACH 1: Two-pass Linear DP (OPTIMAL)
 *
 * Key Insight:
 * Since the houses form a circle, house[0] and house[n-1] are adjacent.
 * Break the circle by solving two linear sub-problems:
 *   - Rob houses 0..n-2 (exclude last)
 *   - Rob houses 1..n-1 (exclude first)
 * Answer = max of both sub-problems.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
class Solution1 {
private:
    int robLinear(vector<int>& nums, int lo, int hi) {
        int prev2 = 0, prev1 = 0;
        for (int i = lo; i <= hi; i++) {
            int cur = max(prev1, prev2 + nums[i]);
            prev2 = prev1;
            prev1 = cur;
        }
        return prev1;
    }

public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if (n == 1) return nums[0];
        if (n == 2) return max(nums[0], nums[1]);
        return max(robLinear(nums, 0, n-2),
                   robLinear(nums, 1, n-1));
    }
};

/*
 * APPROACH 2: DP with Array (Two-pass, explicit dp arrays)
 *
 * Same two-pass idea but store full dp[] to trace the path.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */
class Solution2 {
private:
    int robRange(vector<int>& nums, int lo, int hi) {
        if (lo == hi) return nums[lo];
        vector<int> dp(nums.size(), 0);
        dp[lo] = nums[lo];
        dp[lo+1] = max(nums[lo], nums[lo+1]);
        for (int i = lo+2; i <= hi; i++)
            dp[i] = max(nums[i] + dp[i-2], dp[i-1]);
        return dp[hi];
    }

public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if (n == 1) return nums[0];
        if (n == 2) return max(nums[0], nums[1]);
        return max(robRange(nums, 0, n-2),
                   robRange(nums, 1, n-1));
    }
};

/*
 * APPROACH 3: Memoization (Top-Down)
 *
 * Two-pass top-down DP with memoization.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */
class Solution3 {
private:
    vector<int> memo;

    int helper(vector<int>& nums, int i, int hi) {
        if (i > hi) return 0;
        if (memo[i] != -1) return memo[i];
        memo[i] = max(nums[i] + helper(nums, i+2, hi),
                      helper(nums, i+1, hi));
        return memo[i];
    }

public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if (n == 1) return nums[0];
        if (n == 2) return max(nums[0], nums[1]);

        memo.assign(n, -1);
        int res1 = helper(nums, 0, n-2);

        memo.assign(n, -1);
        int res2 = helper(nums, 1, n-1);

        return max(res1, res2);
    }
};

/*
 * APPROACH 4: Rob/Skip state machine (Two-pass)
 *
 * Uses explicit rob/skip states (same as House Robber I Approach 4)
 * applied twice over the two ranges.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
class Solution4 {
private:
    int robRange(vector<int>& nums, int lo, int hi) {
        int rob = 0, skip = 0;
        for (int i = lo; i <= hi; i++) {
            int newRob  = skip + nums[i];
            int newSkip = max(rob, skip);
            rob  = newRob;
            skip = newSkip;
        }
        return max(rob, skip);
    }

public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if (n == 1) return nums[0];
        if (n == 2) return max(nums[0], nums[1]);
        return max(robRange(nums, 0, n-2),
                   robRange(nums, 1, n-1));
    }
};

// ==================== TEST CASES ====================
void runTests() {
    Solution1 s1; Solution2 s2; Solution3 s3; Solution4 s4;

    auto run = [&](vector<int> nums, int expected) {
        cout << "Input: [";
        for (int i = 0; i < (int)nums.size(); i++) {
            cout << nums[i]; if (i+1 < (int)nums.size()) cout << ",";
        }
        cout << "]\n";
        cout << "  S1=" << s1.rob(nums)
             << "  S2=" << s2.rob(nums)
             << "  S3=" << s3.rob(nums)
             << "  S4=" << s4.rob(nums)
             << "  Expected=" << expected << "\n\n";
    };

    run({2,3,2},     3);
    run({1,2,3,1},   4);
    run({1,2,3},     3);
    run({1},         1);
    run({2,1},       2);
    run({5,5,5,5,5}, 15);
}

int main() {
    runTests();
    return 0;
}

/*
 * COMPLEXITY ANALYSIS SUMMARY:
 *
 * Approach 1 (Two-pass O(1) space — RECOMMENDED):
 *   Time: O(n), Space: O(1)
 *
 * Approach 2 (Two-pass with dp array):
 *   Time: O(n), Space: O(n)
 *
 * Approach 3 (Memoization top-down):
 *   Time: O(n), Space: O(n)
 *
 * Approach 4 (Rob/Skip state machine):
 *   Time: O(n), Space: O(1)
 *
 * KEY INSIGHTS:
 * 1. Circular constraint → break into two linear sub-problems
 * 2. Sub-problem 1: rob houses [0, n-2] (first included, last excluded)
 * 3. Sub-problem 2: rob houses [1, n-1] (first excluded, last included)
 * 4. Answer = max(sub1, sub2)
 * 5. Each sub-problem is identical to House Robber I
 *
 * COMMON MISTAKES:
 * - Forgetting that first and last are adjacent (circular)
 * - Only doing one pass instead of two
 * - Incorrect range boundaries [0,n-2] vs [1,n-1]
 */

// Made with Bob

File: src/data/tutorials/053-house-robber-ii.js
/**
 * Tutorial content for #053 — House Robber II
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `You are a robber and the houses are arranged in a circle — the first house and the last house are neighbours. You cannot rob two adjacent houses. Given an array of house values, return the maximum amount you can rob tonight without triggering the alarm.`,
    example: `houses = [2, 3, 2]\n→ Cannot rob house 0 and house 2 together (they are adjacent in the circle)\n→ Best: rob house 1 alone = 3\n✅ Answer: 3`,
    keyInsight: `Because the street is circular, robbing the first house prevents robbing the last. Break the circle into two linear sub-problems: houses 0..n-2 (skip last) and houses 1..n-1 (skip first). The answer is the max of both.`,
  },

  approaches: {
    'Two-Pass Linear DP': {
      intuition: `The circular constraint means first and last can't both be robbed. So solve two independent linear House Robber I problems: one that excludes the last house, one that excludes the first. Take the maximum. Each pass uses the O(1)-space sliding-variable trick.`,
      steps: [
        `Handle base cases: n=1 → return nums[0]; n=2 → return max(nums[0], nums[1]).`,
        `Run robRange(nums, 0, n-2): the linear robber on houses 0 to n-2.`,
        `Run robRange(nums, 1, n-1): the linear robber on houses 1 to n-1.`,
        `Return the maximum of the two results.`,
        `Inside robRange: use prev2, prev1 sliding variables; cur = max(prev1, prev2 + nums[i]).`,
      ],
      example: `houses = [2, 3, 2]\n\nPass 1 (indices 0..1): houses [2, 3]\n  prev2=0, prev1=0\n  i=0: cur=max(0,0+2)=2 → prev2=0, prev1=2\n  i=1: cur=max(2,0+3)=3 → result1=3\n\nPass 2 (indices 1..2): houses [3, 2]\n  i=1: cur=max(0,0+3)=3 → prev2=0, prev1=3\n  i=2: cur=max(3,0+2)=3 → result2=3\n\nmax(3, 3) = 3\n✅ Answer: 3`,
      keyInsight: `O(n) time and O(1) space. The two-pass decomposition is the standard interview answer for any circular DP variant.`,
    },

    'Explicit Two-Pass with Result Array': {
      intuition: `Same two-pass idea, but instead of O(1) variables we allocate full dp arrays for each pass. This is more verbose but lets you inspect every intermediate value — useful for debugging or visualisation.`,
      steps: [
        `Handle base cases for n=1 and n=2.`,
        `Build dp1[0..n-2]: dp1[0]=nums[0], dp1[1]=max(nums[0],nums[1]), then dp1[i]=max(dp1[i-1], dp1[i-2]+nums[i]).`,
        `Build dp2[1..n-1]: dp2[1]=nums[1], dp2[2]=max(nums[1],nums[2]), then dp2[i]=max(dp2[i-1], dp2[i-2]+nums[i]).`,
        `Return max(dp1[n-2], dp2[n-1]).`,
      ],
      example: `houses = [2, 3, 2]  (n=3)\n\ndp1 (indices 0..1):\n  dp1[0]=2, dp1[1]=max(2,3)=3\n\ndp2 (indices 1..2):\n  dp2[1]=3, dp2[2]=max(3,2)=3\n\nmax(dp1[1], dp2[2]) = max(3, 3) = 3\n✅ Answer: 3`,
      keyInsight: `O(n) time and O(n) space. Use this when you need to trace or display the dp table; use Two-Pass Linear DP in production for O(1) space.`,
    },

    'Memoization (Top-Down)': {
      intuition: `Solve each range recursively with memoisation. Two separate memo arrays guard the two subproblems: [0..n-2] and [1..n-1]. At each index, choose whether to rob the current house (skip two ahead) or skip it (advance one).`,
      steps: [
        `Handle base cases for n=1 and n=2.`,
        `Create memo1 and memo2, both size n, filled with -1.`,
        `Call robMemoHelper(nums, 0, n-2, memo1) for the first range.`,
        `Call robMemoHelper(nums, 1, n-1, memo2) for the second range.`,
        `Inside helper: if i > hi return 0; if memo[i] != -1 return it; else memo[i] = max(nums[i] + helper(i+2), helper(i+1)).`,
        `Return max of both helper results.`,
      ],
      example: `houses = [2, 3, 2]\n\nRange [0..1]:\n  helper(0): max(2+helper(2), helper(1))\n    helper(2): 2>1 → 0\n    helper(1): max(3+helper(3), helper(2)) = max(3+0,0) = 3\n  helper(0) = max(2+0, 3) = 3\n\nRange [1..2]:\n  helper(1): max(3+helper(3), helper(2))\n    helper(3): 3>2 → 0\n    helper(2): max(2+helper(4), helper(3)) = 2\n  helper(1) = max(3+0, 2) = 3\n\nmax(3, 3) = 3\n✅ Answer: 3`,
      keyInsight: `O(n) time and O(n) space. Top-down style is natural if you think recursively; the memo arrays avoid redundant re-computation.`,
    },

    'Rob/Skip State Machine (Two-Pass)': {
      intuition: `Model each pass as an explicit rob/skip state machine. At each house, robVal is the best total if you rob this house (so you must have skipped the previous), and skipVal is the best if you skip (free to do either next). Run this machine twice over the two sub-ranges and take the max.`,
      steps: [
        `Handle base cases for n=1 and n=2.`,
        `For each range, start robVal=0, skipVal=0.`,
        `For each house i in the range: newRob = skipVal + nums[i]; newSkip = max(robVal, skipVal).`,
        `Update robVal = newRob, skipVal = newSkip.`,
        `Return max(robVal, skipVal) for the range.`,
        `Overall answer = max of the two range results.`,
      ],
      example: `houses = [2, 3, 2]\n\nRange [0..1]: houses [2, 3]\n  rob=0, skip=0\n  i=0 (2): newRob=0+2=2, newSkip=max(0,0)=0 → rob=2, skip=0\n  i=1 (3): newRob=0+3=3, newSkip=max(2,0)=2 → rob=3, skip=2\n  result1 = max(3,2) = 3\n\nRange [1..2]: houses [3, 2]\n  rob=0, skip=0\n  i=1 (3): rob=3, skip=0\n  i=2 (2): newRob=0+2=2, newSkip=max(3,0)=3 → rob=2, skip=3\n  result2 = max(2,3) = 3\n\nmax(3, 3) = 3\n✅ Answer: 3`,
      keyInsight: `O(n) time and O(1) space. The state-machine framing makes the adjacency constraint crystal-clear: you can only enter a "rob" state from a "skip" state.`,
    },
  },
}
